nntrainer/layers/pooling2d_layer.cpp

```diff
@@ -40,25 +41,55 @@ int Pooling2DLayer::initialize(Manager &manager) {
     ml_logw("Warning: the length of previous layer dimension is one");
   }
 
+  if (pooling_type == PoolingType::global_max ||
+      pooling_type == PoolingType::global_average) {
+    if (pool_size[0] != in_dim.height() || pool_size[1] != in_dim.width()) {
+      ml_logw(
+        "[Pooling2D] global_max, global_average does not except pool size");
+      pool_size[0] = in_dim.height();
+      pool_size[1] = in_dim.width();
+    }
+
+    if (padding[0] != 0 || padding[1] != 0) {
+      ml_logw("[Pooling2D] global_max, global_average does not except padding");
+      padding[0] = 0;
+      padding[1] = 0;
+    }
+
+    if (stride[0] != 1 || stride[1] != 1) {
+      ml_logw("[Pooling2D] global_max, global_average does not except stride");
+      stride[0] = 1;
+      stride[1] = 1;
+    }
+  }
+
+  unsigned int eff_in_height = in_dim.height() + padding[0] * 2;
+  unsigned int eff_in_width = in_dim.width() + padding[1] * 2;
+
+  if (eff_in_height < pool_size[0] || eff_in_width < pool_size[1]) {
+    ml_loge("[Pooling2D] Failed to initialize: in size + padding is smaller "
+            "than effective kernel");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  unsigned int IM = std::numeric_limits<int>::max();
+
+  if (eff_in_height - padding[0] - pool_size[0] > IM ||
+      eff_in_width - padding[1] - pool_size[1] > IM) {
+    ml_loge(
+      "[Pooling2D] Failed to initialize: Calculated patch end is over int max");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
   out_dim.batch(in_dim.batch());
   out_dim.channel(in_dim.channel());
+  out_dim.height((eff_in_height - pool_size[0]) / stride[0] + 1);
+  out_dim.width((eff_in_width - pool_size[1]) / stride[1] + 1);
 
-  if (pooling_type == PoolingType::max ||
-      pooling_type == PoolingType::average) {
-    out_dim.height(
-      (in_dim.height() - pool_size[0] + 2 * padding[0]) / stride[0] + 1);
-    out_dim.width((in_dim.width() - pool_size[1] + 2 * padding[1]) / stride[1] +
-                  1);
+  if (pooling_type == PoolingType::global_max) {
+    max_idx_global.resize(out_dim.batch() * out_dim.getFeatureLen());
   } else {
-    out_dim.height(1);
-    out_dim.channel(1);
-    out_dim.width(in_dim.channel());
-  }
-
-  if (pooling_type == PoolingType::max) {
-    max_idx.reserve(in_dim.batch() * output_dim[0].getFeatureLen());
-  } else if (pooling_type == PoolingType::global_max) {
-    max_idx_global.resize(out_dim.batch() * output_dim[0].getFeatureLen());
+    max_idx.reserve(in_dim.batch() * out_dim.getFeatureLen());
```

Comment:
i wonder why the behavior of max_idx and max_idx_global is different (one resizes while the other reserves)?

Reply:
max_idx_global will move to reserve in the upcoming PR. global_max_pooling is left untouched in this version.